python/src/aac/plugins/validators/unused_definitions/_validate_used_definitions.py

```diff
@@ -0,0 +1,57 @@
+import logging
+
+from aac.lang.definitions.definition import Definition
+from aac.lang.definitions.references import get_definition_type_references_from_list
+from aac.lang.language_context import LanguageContext
+from aac.plugins.validators import ValidatorFindings, ValidatorResult
+
+
+PLUGIN_NAME = "Validate definition references"
+VALIDATION_NAME = "Definition is referenced"
+
+
+def validate_used_definitions(
+    definition_under_test: Definition,
+    target_schema_definition: Definition,
+    language_context: LanguageContext,
+    *validation_args,
+) -> ValidatorResult:
+    """
+    Validates that the definition has valid definition references.
+
+    Args:
+        definition_under_test (Definition): The definition that's being validated.
+        target_schema_definition (Definition): The schema definition with the validation rules that trigger the validation.
+        language_context (LanguageContext): A management and utility classfor the contextual AaC domain-specific language.
+
+    Returns:
+        A ValidatorResult containing any applicable error messages.
+    """
+    findings = ValidatorFindings()
+
+    # Check that definition to test does exist within the language context
+    if language_context.get_definition_by_name(definition_under_test.name):
+
+        # Check for additional references within the language context of the tested defintion
+        referenced_definitions = get_definition_type_references_from_list(definition_under_test, language_context.definitions)
+
+        # Get lexme for contributing to findings messages
+        reference_lexeme = definition_under_test.get_lexeme_with_value(definition_under_test.name)
+
+        if len(referenced_definitions) == 0:
+            no_references_found_message = f"No references to '{definition_under_test.name}' in the language context."
+            logging.info(no_references_found_message)
+            if reference_lexeme:
+                findings.add_info_finding(
+                    definition_under_test, no_references_found_message, PLUGIN_NAME, reference_lexeme
+                )
+    else:
+        no_definition_message = f"No definition with the name '{definition_under_test.name}' was found in the language context."
+        logging.error(no_definition_message)
+
+        if reference_lexeme:
+            findings.add_error_finding(
+                definition_under_test, no_definition_message, PLUGIN_NAME, reference_lexeme
+            )
```

Comment:
The remaining failing tests are returning issues where line 52 `if reference_lexeme:` is referencing the variable `reference_lexeme` that is only assigned in the if/true scope on line 39 `reference_lexeme = definition_under_test.get_lexeme_with_value(definition_under_test.name)`. I tried moving the assignment to outside the if/true scope so that it was present in the else scope, but the one remaining test still fails as it's now getting an error finding that the test definition isn't in the context. 
https://github.com/jondavid-black/AaC/blob/71c6aaedd68e1b825805015f41702521ac2dc8b9/python/tests/validate/test_validate_definition.py#L28



When we run validation we add all of the definitions being validated to the context prior to validating them. https://github.com/jondavid-black/AaC/blob/71c6aaedd68e1b825805015f41702521ac2dc8b9/python/src/aac/validate/_validate.py#L88-L110

The only instances when the definition won't be in the context is when we run these unit tests directly on validator functions. Definitions not being in the context is only an issue if they're referenced by other definitions and we need to do some lookup to get the referenced definition. In this particular validation we're not following references, we're just checking for a non-zero number of references to the definition being validated by other definitions in the context. I'd be happy to elaborate more during standup, but it's not really necessary for us to perform the check that the definition being validated is in the context -- and it's giving this validator multiple behaviors which we try to avoid with validators.

Reply:
Generated issue #639 as discussed for re-evaluating this impact.